File: Day11_Maximum Product Subarray/maximum_prod_subarray.py
# https://www.geeksforgeeks.org/batch/gfg-160-problems/track/arrays-gfg-160/problem/maximum-product-subarray3604

def maxProduct(arr):
    # code here
    prefix, suffix = 1,1
    maxi = float("-inf")
    n = len(arr)
    for i in range(n): # check prefix max and suffix max 
        if prefix == 0:
            prefix = 1
        if suffix == 0:
            suffix = 1
        prefix *= arr[i]
        suffix *= arr[n-i-1]
        maxi = max(maxi,prefix,suffix)
    return maxi

user_input = input("Enter space-separated integers: ")
arr = list(map(int, user_input.split()))
print(maxProduct(arr=arr))

File: Day11_Maximum Product Subarray/maximum_prod_subarray.java
import java.util.ArrayList;
import java.util.Scanner;

public class maximum_prod_subarray {
    public static int maxProduct(ArrayList<Integer> arr) { 
        int prefix = 1;
        int suffix = 1;
        int result = Integer.MIN_VALUE;
        int n = arr.size();
        for (int i = 0; i < n; i++) {
            if(prefix==0)
                prefix = 1;
            if(suffix==0)
                suffix = 1;
            prefix *=  arr.get(i);
            suffix *= arr.get(n-i-1);
            result = Math.max(result,Math.max(prefix,suffix));
        }
        return result;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size and element in array: ");
        int n = sc.nextInt();
        ArrayList<Integer> arr = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            arr.add(sc.nextInt());
        }
        int result = maxProduct(arr);
        System.out.println("Maximum product: "+result);
        sc.close();
    }
}
